miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFileDownloader.kt

```diff
@@ -152,4 +178,50 @@ class MiniAppFileDownloaderDefault(var activity: Activity, var requestCode: Int)
         }
         return builder.url(url).build()
     }
+
+    private val String.isHttpUrl get() = this.startsWith("https:") || this.startsWith("http:")
+    private val String.isDataUri get() = this.startsWith("data:")
+    private val String.isValidUrl get() = this.isHttpUrl || this.isDataUri
+
+    private fun saveDataUri(
+        destinationUri: Uri,
+        url: String,
+        fileName: String,
+        onDownloadSuccess: (String) -> Unit,
+        onDownloadFailed: (MiniAppDownloadFileError) -> Unit
+    ) {
+        try {
+            val decodedBytes = Base64.decode(
+                url.substring(url.indexOf(",") + 1),
+                Base64.DEFAULT
+            )
+            saveFile(
+                destinationUri = destinationUri,
+                data = decodedBytes.inputStream(),
+                fileName = fileName,
+                onDownloadSuccess = onDownloadSuccess,
+                onDownloadFailed = onDownloadFailed
+            )
+        } catch (e: IllegalArgumentException) {
+            Log.e("MiniAppFileDownloader", "Failed to download file: could not decode URI as a Base64 string.")
+            onDownloadFailed.invoke(MiniAppDownloadFileError.saveFailureError)
+        }
+    }
+
+    private fun saveFile(
+        destinationUri: Uri,
+        data: InputStream,
+        fileName: String,
+        onDownloadSuccess: (String) -> Unit,
+        onDownloadFailed: (MiniAppDownloadFileError) -> Unit
+    ) {
+        activity.contentResolver.openOutputStream(destinationUri)?.use { outputStream ->
+            data.copyTo(outputStream)
+            outputStream.close()
+            onDownloadSuccess.invoke(fileName)
+        } ?: run {
+            onDownloadFailed.invoke(MiniAppDownloadFileError.saveFailureError)
+            Log.e("MiniAppFileDownloader", "Failed to download file: could not open OutputStream.")
```

Comment:
Same as above.
I think this `Log.e` should be removed if it's logging locally but if it's logging to firebase then may be the error message could `Failed to download file: Error occurred while opening the OutputStream.

Reply:
Thanks, I updated this.